File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int[][] board = new int[10][10];
    static int[] remain = {0, 5, 5, 5, 5, 5};
    static int result = Integer.MAX_VALUE;

    private static boolean isValid(int x, int y, int size) {
        if (x + size > 10 || y + size > 10) return false;

        for (int i = y; i < y + size; i++) {
            for (int j = x; j < x + size; j++) {
                if (board[i][j] == 0) return false;
            }
        }

        return true;
    }

    private static void fill(int x, int y, int size, int flag) {
        for (int i = y; i < y + size; i++) {
            for (int j = x; j < x + size; j++) {
                board[i][j] = flag;
            }
        }
    }

    private static void backtracking(int x, int y, int count) {
        if (count >= result) return;

        if (y == 10) {
            result = count;
            return;
        }

        int nextX = (x + 1) % 10;
        int nextY = (nextX == 0) ? y + 1 : y;

        if (board[y][x] == 0) {
            backtracking(nextX, nextY, count);
            return;
        }

        for (int size = 5; size >= 1; size--) {
            if (remain[size] == 0 || !isValid(x, y, size)) continue;

            remain[size]--;
            fill(x, y, size, 0);
            backtracking(nextX, nextY, count + 1);
            fill(x, y, size, 1);
            remain[size]++;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        for (int i = 0; i < 10; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 10; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        backtracking(0, 0, 0);

        if (result == Integer.MAX_VALUE) {
            System.out.println(-1);
        } else {
            System.out.println(result);
        }
    }
}

File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

val board = Array(10) { IntArray(10) }
val remain = intArrayOf(0, 5, 5, 5, 5, 5)
var result = Int.MAX_VALUE

fun isValid(x: Int, y: Int, size: Int): Boolean {
    if (x + size > 10 || y + size > 10) return false

    for (i in y until y + size) {
        for (j in x until x + size) {
            if (board[i][j] == 0) return false
        }
    }

    return true
}

fun fill(x: Int, y: Int, size: Int, flag: Int) {
    for (i in y until y + size) {
        for (j in x until x + size) {
            board[i][j] = flag
        }
    }
}

fun backtracking(x: Int, y: Int, count: Int) {
    if (count >= result) return

    if (y == 10) {
        result = count
        return
    }

    val nextX = (x + 1) % 10
    val nextY = if (nextX == 0) y + 1 else y

    if (board[y][x] == 0) {
        backtracking(nextX, nextY, count)
        return
    }

    for (size in 5 downTo 1) {
        if (remain[size] == 0 || !isValid(x, y, size)) continue

        remain[size]--
        fill(x, y, size, 0)
        backtracking(nextX, nextY, count + 1)
        fill(x, y, size, 1)
        remain[size]++
    }
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    for (i in 0 until 10) {
        val st = StringTokenizer(br.readLine())
        for (j in 0 until 10) {
            board[i][j] = st.nextToken().toInt()
        }
    }

    backtracking(0, 0, 0)

    println(if (result == Int.MAX_VALUE) -1 else result)
}